ee/kernel/src/timer.c

```diff
@@ -13,7 +13,1179 @@
  * Some routines to do some timer work
  */
 
-#include <tamtypes.h>
+#include <kernel.h>
+#include <timer.h>
+#include <string.h>
+#include <stdint.h>
+
+#ifdef F_timer_data
+timer_ee_global_struct g_Timer;
+timer_ee_entry_struct g_CounterBuf[128];
+alarm_ee_entry_struct g_AlarmBuf[64];
+alarm_ee_entry_struct *g_pFreeAlarm = NULL;
+#else
+extern timer_ee_global_struct g_Timer;
+extern timer_ee_entry_struct g_CounterBuf[128];
+extern alarm_ee_entry_struct g_AlarmBuf[64];
+extern alarm_ee_entry_struct *g_pFreeAlarm;
+#endif
+
+#ifdef F_SetT2
+s32 SetT2(volatile void *a1, u32 a2)
+{
+	u32 v2;
+	u32 v3;
+	u32 v5;
+
+	__asm__ __volatile__ ("mfc0\t%0, $12" : "=r" (v3));
+	v3 = v3 & 0x10000;
+	if (v3 != 0)
+	{
+		do
+		{
+			DI();
+			EE_SYNCP();
+			__asm__ __volatile__ ("mfc0\t%0, $12" : "=r" (v5));
+		}
+		while ((v5 & 0x10000) != 0);
+	}
+	__asm__ __volatile__ ("mfc0\t%0, $12" : "=r" (v5));
+	EE_SYNCP();
+	v5 = ((v5 | 6) ^ 2) | v3;
+	__asm__ __volatile__ ("mtc0\t%0, $12" : "=r" (v5));
+	EE_SYNCP();
+	*((volatile u32 *)a1) = a2;
+	__asm__ __volatile__ ("mtc0\t%0, $30" : "=r" (v2));
+	EE_SYNCP();
+	__asm__ __volatile__ ("eret");
+	return v2;
+}
+#endif
+
+#ifdef F_SetT2_COUNT
+void SetT2_COUNT(u32 a1)
+{
+	SetT2((volatile void *)0xB0001000, a1);
+}
+#endif
+
+#ifdef F_SetT2_MODE
+void SetT2_MODE(u32 a1)
+{
+	SetT2((volatile void *)0xB0001010, a1);
+}
+#endif
+
+#ifdef F_SetT2_COMP
+void SetT2_COMP(u32 a1)
+{
+	SetT2((volatile void *)0xB0001020, a1);
+}
+#endif
+
+#ifdef F_InitTimer
+s32 InitTimer(s32 a1)
+{
+	s32 v6;
+	u32 oldintr;
+	u32 v8;
+
+	if (g_Timer.intc_handler >= 0)
```

Comment:
I think that this comparison should be `if (g_Timer.intc_handler > 0)` otherwise the timer won't do the init properly because by default the `g_Timer` is empty

Reply:
Or the initial value of `g_Timer` is wrong defined